tests/cuda/matrix_vector_mf.cuh

```diff
@@ -0,0 +1,118 @@
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+// This is a template for matrix-vector products with the Helmholtz equation
+// (zero and first derivatives) on different kinds of meshes (Cartesian,
+// general, with and without hanging nodes).
+
+#include "../tests.h"
+
+#include <deal.II/matrix_free/cuda_matrix_free.h>
+#include <deal.II/matrix_free/cuda_fe_evaluation.cuh>
+
+#include <deal.II/lac/cuda_vector.h>
+
+template <int dim, int fe_degree, typename Number, int n_q_points_1d>
+class HelmholtzOperatorQuad
+{
+public :
+  __device__ void operator()(
+    CUDAWrappers::FEEvaluation<dim,fe_degree,n_q_points_1d,1,Number> *fe_eval,
+    const unsigned int q_point) const;
+};
+
+
+
+template <int dim, int fe_degree, typename Number, int n_q_points_1d>
+__device__ void
+HelmholtzOperatorQuad<dim,fe_degree,Number,n_q_points_1d>::operator()(
+  CUDAWrappers::FEEvaluation<dim,fe_degree,n_q_points_1d,1,Number> *fe_eval,
+  const unsigned int q) const
+{
+  fe_eval->submit_value(Number(10)*fe_eval->get_value(q),q);
+  fe_eval->submit_gradient(fe_eval->get_gradient(q),q);
+}
+
+
+
+template <int dim, int fe_degree, typename Number, int n_q_points_1d>
+class HelmholtzOperator
+{
+public:
+  __device__ void operator()(
+    const unsigned int cell,
+    const typename CUDAWrappers::MatrixFree<dim,Number>::Data *gpu_data,
+    CUDAWrappers::SharedData<dim,Number>                *shared_data,
+    const Number                                        *src,
+    Number                                              *dst) const;
+
+  static const unsigned int n_dofs_1d = fe_degree+1;
+  static const unsigned int n_local_dofs =
+    dealii::Utilities::fixed_int_power<fe_degree+1,dim>::value;
+  static const unsigned int n_q_points =
+    dealii::Utilities::fixed_int_power<n_q_points_1d,dim>::value;
+};
+
+
+
+template <int dim, int fe_degree, typename Number, int n_q_points_1d>
+__device__ void
+HelmholtzOperator<dim,fe_degree,Number,n_q_points_1d>::operator()(
+  const unsigned int                                        cell,
+  const typename CUDAWrappers::MatrixFree<dim,Number>::Data *gpu_data,
+  CUDAWrappers::SharedData<dim,Number>                *shared_data,
+  const Number                                        *src,
+  Number                                              *dst) const
+{
+  CUDAWrappers::FEEvaluation<dim,fe_degree,n_q_points_1d,1,Number> fe_eval(
+    cell, gpu_data, shared_data);
+  fe_eval.read_dof_values(src);
+  fe_eval.evaluate(true, true);
+  fe_eval.apply_quad_point_operations(
+    HelmholtzOperatorQuad<dim,fe_degree,Number,n_q_points_1d>());
+  fe_eval.integrate(true, true);
+  fe_eval.distribute_local_to_global (dst);
```

Comment:
@kalj Yes we could pass that down together with the functor, possibly also as a (static) member variable, for the moment.

> but are we sure it is not too restrictive?

Well, it certainly is too restrictive in some cases, so there should be some way to get access to this more verbose form. But it does not have to be as pretty. Applications of that kind are for example settings where you compute a stabilization term as the maximum over some quantities on all quadrature points, as e.g. in step-31/step-32.